### DISCUSSION TITLE
[Frontend] Compose Multiplatform + Navigation 3 + MVIKotlin

### DISCUSSION THREAD

**Discussion by CXwudi** [realworld-apps/realworld#1545] [WIP Implementations]
2026.09 Last Update: Project is migrated to use Navigation 3 instead of Decompose.

2025.10 Update: With the minimal set of feature in frontend completed, I am now considering this project as completed, and move on to the next practice project. I will still keep updating dependencies and keep the project deprecation-free.

2024.10 Update: I am slowly continuing the project with a minimal set of features in frontend first

2024.08 Update: I am pausing the implementation for now. More information [below](https://github.com/gothinkster/realworld/discussions/1545#discussioncomment-10165234)

---

I am planning to start this practice at https://github.com/CXwudi/realworld-compose-http4k-example-app, where I will be using [Compose Multiplatform](https://www.jetbrains.com/lp/compose-multiplatform/) as the frontend and [http4k](https://www.http4k.org/) as the backend. All the code will be written in Kotlin, and common code will be shared using [Kotlin Multiplatform](https://kotlinlang.org/docs/multiplatform.html) technology.

In case you're wondering, Compose Multiplatform is JetBrains' adaptation from the original Android Jetpack Compose framework, which is highly inspired by other declarative UI frameworks like React and Flutter. We have several Android frontend examples in [codebase.show](https://codebase.show/projects/realworld?category=frontend&language=kotlin), but none of them use Compose. Compose Multiplatform supports Android, Desktop (JVM), iOS, and Web. I can support all platforms but except iOS, because I don't have a Mac machine. I can still try to make the iOS frontend compile successfully in my MacOS VM, which can't launch the iPhone simulator.

We already have a backend example written using http4k, but I would like to implement mine with the [Koin](https://insert-koin.io/) DI framework. Also, the backend will be used to demonstrate the code-sharing capability from Kotlin Multiplatform, where common codes (mostly just POJOs with the Kotlin serialization plugin) are shared between the frontend and backend.

Since the UI will be written in Compose Multiplatform, I won't be able to reuse the theme and HTML code snippet in the frontend spec. Instead, I will follow the [mobile spec](https://github.com/gothinkster/realworld/blob/main/apps/documentation/docs/specs/mobile-specs/introduction.md) and take some UI design ideas from the Medium app.

Lastly, the development scope is quite huge. As a new grad, it might take me a huge effort and time to complete it. Or, in the worst case, it won't be finished. Nevertheless, I will keep working on it to improve myself, unless I get a job offer or a master's degree offer.

**Comment by CXwudi** [comment]
2024.08 Update: After a year of dedicated job searching, I've made the strategic decision to invest in my professional development through an intensive IT training program. As a result, I will be temporarily stepping away from this project to focus entirely on enhancing my skills and following the program's guidance. 

I've thoroughly enjoyed the time spent solving challenges, learning, and experimenting with [Compose Multiplatform](https://www.jetbrains.com/lp/compose-multiplatform/), [Decompose](https://github.com/arkivanov/Decompose), and [MVIKotlin](https://github.com/arkivanov/MVIKotlin). I'm eagerly looking forward to the day when I can return to this project.

Meanwhile, If you're interested in taking over this project or starting a new project with a similar tech stack, please feel free to do so. I wouldn't want this project to discourage others from trying the similar ideas.

**Comment by CXwudi** [comment]
2024.10 Update: The intensive IT training is now in a stable stage. Therefore, I am now continuing the project. However, due to the complexity of [Compose Multiplatform](https://www.jetbrains.com/lp/compose-multiplatform/), [Decompose](https://github.com/arkivanov/Decompose), and [MVIKotlin](https://github.com/arkivanov/MVIKotlin), this project will be scaled down to only [a minimal set of features](https://github.com/CXwudi/realworld-compose-http4k-example-app/issues/21), as suggested in the [above comment](https://github.com/gothinkster/realworld/discussions/1545#discussioncomment-9899294). Additionally, I will be focusing only on the frontend part for now, and I will only continue with the backend part if time allows.

For now, I will not remove the "Discontinued" label from the title, as it is uncertain whether I will need more time for IT training.

**Comment by CXwudi** [comment]
2024.11 Update: I finally landed a job. Originally I would abandon this project upon landing a job, but I decided to continue the minimal implementation mentioned above.

**Comment by CXwudi** [comment]
2025.10 Update: With the [minimal set of feature](https://github.com/CXwudi/realworld-compose-http4k-example-app/issues/21) completed, I am now considering my project as completed

**Comment by CXwudi** [comment]
Keep this discussion as Open to let it searchable

**Comment by c4ffein** [comment]
Nice, you can submit this at https://codebase.show/projects/realworld/implementations/submit if you haven’t already :)

**Reply by CXwudi** [reply]
Hi, my submission was rejected, maybe is it because my repo name was confusing because it was a fullstack project before? I have renamed the repo https://github.com/CXwudi/realworld-cmp-decompose-mvikotlin-app to properly reflect that this is now a frontend project

I have submitted a new one with renamed repo and updated description

**Reply by c4ffein** [reply]
@mvila if you want to explain what’s wrong

**Reply by mvila** [reply]
I'm very sorry, @CXwudi, but we can't approve an implementation that doesn't include the required features. If this was allowed in the past, it was a mistake.

**Comment by CXwudi** [comment]
Last Update: 2026.09.19

With Navigation 3 is now stable and recommended navigation for CMP app, I (with GPT-6 Astra xhigh) migrate from using Decompose to using official AndroidX ViewModel and Naviagtion 3. Tho the MVIKotlin is kept.

Since this post was originally written for Decompose, so I now closing it as nowadays it is not recommended to use any other navigation solution other than Navigation 3
